### ISSUE TITLE
Request/question: Slow QuakeML reading

### ISSUE THREAD

**Issue by calum-chamberlain** [obspy/obspy#1910]
QuakeML reading is quite slow - in my [naive test](https://gist.github.com/calum-chamberlain/a455a6f38a8dc6360ed0e52498d7799e#file-obspy_event_io-ipynb), about 10x slower than writing. For large catalog files this becomes a bit of an issue.

The main point of this issue is to ask:
1. Is there a faster IO alternative already implemented?
2. It looks to me like the main slow-down is the loop over events in obspy.io.quakeml.core.Unpickler._deserialize, I wonder if this can be a threaded operation?
3. If none of the above, does anyone have any ideas for speeding up the reading of QuakeML files to ObsPy Events?

If any suggestions are made I'm happy to undertake the work to do it.

**Comment by megies** [comment]
> Is there a faster IO alternative already implemented?

Not to my knowledge.



> It looks to me like the main slow-down is the loop over events in obspy.io.quakeml.core.Unpickler._deserialize, I wonder if this can be a threaded operation?

Sounds like a pretty good idea. The FDSN `client.py` has threading for the individual endpoints and might be a good starting point for this.

**Comment by calum-chamberlain** [comment]
Thanks for that @megies, I will have a look there and crack on with trying to speed it up.

**Comment by megies** [comment]
No idea if there's hidden pitfalls I don't see right now with multiple threads working on the xml object, but I guess it shouldn't be too much work to set it up and see what happens..

**Comment by d-chambers** [comment]
Hey @calum-chamberlain,

If you haven't used it, I recommend checking out [snakeviz](https://jiffyclub.github.io/snakeviz/), it makes profiling things like this a delightful experience. 

I did something similar trying to speed up reading miniseed files (https://github.com/obspy/obspy/pull/1845) and was able to find some low hanging fruit very quickly. [This notebook](https://gist.github.com/d-chambers/1eabd9c2147b11db0e68cf57ac3a8ad1) is a simple example.

**Comment by calum-chamberlain** [comment]
Thanks for those comments guys - I thought I could get around the GIL with more direct access to lxml (which I think releases the GIL?), but it would require some strange in-between stages that I'm not that keen on.

Thanks @d-chambers I haven't used snakeviz - will do and do something similar with a notebook and share results here!

**Comment by calum-chamberlain** [comment]
On that note, while I'm playing around with it, I may have a look at memory usage for writing QuakeML files too - currently it seems like quite an expensive operation - serialising an entire (large) catalog seems to use a lot of additional memory.

**Comment by krischer** [comment]
> Thanks for those comments guys - I thought I could get around the GIL with more direct access to lxml (which I think releases the GIL?), but it would require some strange in-between stages that I'm not that keen on.

Might not be worth it - in the measurements I made above the lxml parsing was less than 1 percent of the full runtime.

**Comment by megies** [comment]
> I did something similar trying to speed up reading miniseed files (#1845) and was able to find some low hanging fruit very quickly. This notebook is a simple example.

Nice! @d-chambers would be great if you link this to https://github.com/obspy/obspy/wiki/Testing%2C-Debugging-%26-Profiling

**Comment by d-chambers** [comment]
Hey @megies, I added a bit in the wiki on this.

**Comment by calum-chamberlain** [comment]
Okay, I have done some timing, and initial playing, with little success. The results are in the gist I posted earlier - re-linked [here](https://gist.github.com/calum-chamberlain/a455a6f38a8dc6360ed0e52498d7799e).
The main point is - calls to `lxml.etree.xpath` take the majority of time.  It is my understanding that we could make use of the XPath class to speed things up a bit there, but probably not by a huge amount, I got about a 3% speed-up by setting up XPath callables for external lookups - that doesn't cover a large space of the xpath calls, but, implementing them all within an iynb is impractical.

Of calls for `obspy.io.quakeml.core._xpath`, about 15% of the time is spent processing `hasattr`, which I think is because `namespace` is not passed to `_xpath`, so we spend time looking up the namespace for each element and subelement of the tree.  Is this necessary? Is there a case where subelements of a QuakeML file have different namespaces to the whole tree? Otherwise, this is a simple speed-up case, but requires changes in lots of places.

Do you think it is worth me going ahead and making those changes in a branch and profiling that? I would probably write my own `Unpickler` outside of obspy first so that I can directly compare the two.

**Comment by megies** [comment]
> Is there a case where subelements of a QuakeML file have different namespaces to the whole tree?

Yes, there are such cases. See http://docs.obspy.org/master/tutorial/code_snippets/quakeml_custom_tags.html?highlight=extra

**Comment by calum-chamberlain** [comment]
Ah, fair enough - will not pursue that strand any further then. I think some more reading on my part of useful ways to read in large xml files needs to be done.
On the note of the `__slots__` attribute, this would probably speed-up creating new objects, but it looks like it would reduce the flexibility of the objects, which would again get in the way of custom attributes for Event subclasses...

**Comment by megies** [comment]
> custom attributes for Event subclasses...

Custom attributes that are handled by QUAKEML I/O are all stored consistently under `extra`. Of course, right now users can add their own custom attributes to event-type objects, though, that can be used in custom workflows.

**Comment by krischer** [comment]
This is a trade-off I would be willing to make to be honest.

**Comment by krischer** [comment]
If the xpath is the slow part: It might be possible to just upfront parse everything to a dictionary and then operate on that with `try`/`except` calls but I'm not sure if that would actually be a lot faster.

**Comment by megies** [comment]
> This is a trade-off I would be willing to make to be honest.

I agree, being able to attach arbitrary attributes to those objects should not be top-priority.

**Comment by d-chambers** [comment]
[Here](https://stackoverflow.com/questions/472000/usage-of-slots) is a good SO question explaining slots in more detail, and [here](https://docs.python.org/3.7/reference/datamodel.html#slots) is the official documentation on the subject for python 3. 

It seems the biggest problem with slots is that it will break multiple inheritance. I am not sure if any users would be sub-classing any of the event objects with multiple inheritance (potentially adding mixins?) but it is not inconceivable.    

Also, if you need dynamic behaviour you can add "\__dict__" to the slots. I would be in favor of doing this as I have several codes that dynamically add methods to these objects, but this is probably not that common.  If you need weak reference support (which I believe the ResourceIdentifiers depend on, so you will) you can add '\__weakref__' to the slots as well.

However, it is doubtful that defining slots will actually speed up object instantiation considerably, as slots is primarily designed to save memory. The instantiating speedups may be even less after defining \__dict__ and \__weakref__. For example, if we take the simplest case:

```python 
# I am using python 3.6.2 on Ubuntu 16.04

class Slots:
   __slots__ = ()

class SlotsDictWeakRef:
    __slots__ = ('__dict__', '__weakref__')

class NoSlots:
    pass

%timeit [Slots() for _ in range(1_000_000)]  # best 160 ms
%timeit [SlotsDictWeakRef() for _ in range(1_000_000)]  # best 182 ms
%timeit [NoSlots() for _ in range(1_000_000)]  # best 206 ms
```
We really don't see much improvement. However, in a more complicated object instantiation the results may vary.

**Comment by megies** [comment]
> event objects with multiple inheritance (potentially adding mixins?) but it is not inconceivable

Actually, I'm doing exactly that in some code of mine (over at megies/obspyck).. ^^

And yeah, slots will probably only do much in terms of memory..

**Comment by calum-chamberlain** [comment]
So after talking with @krischer at SSA, I finally had time to try a different implementation that converts the xml to a dictionary immediately, then makes the objects from the dictionary. This implementation is in [this branch](https://github.com/obspy/obspy/tree/Quakeml-dictionary-reading).  At the moment I haven't finished implementing the `_extra` attributes, but, I think, namespaces stick with the attributes at the moment, so adding that in shouldn't be too difficult. All other tests pass at the moment though (not that that means it is perfect).

This is a really naive implementation of this, and I think it could be done a lot better, but I wanted to see if this made any difference to speed. In a quick and dirty test, reading a 1,000 event catalog (downloaded from GeoNet NZ, so no focal mechanism information) takes **54s** on my machine (average of 7 runs), versus **2min 21s** for the current master implementation (again, average of 7 runs). This is after I removed the `_extra` bit from master so that the comparison was fair.

I'm keen to get opinions on this, and feedback on how this *should* be done if we go down this route, and if anyone sees any obvious issues with doing it this way. I have reused all of the `deserialise` methods, and I haven't touched the `Pickler`.

**Comment by d-chambers** [comment]
Can you also try to pickle and un-pickle the same event on the same machine? I am curious how the un-pickling performance compares.

Going off-topic a bit, it could be beneficial to put together a benchmark suite for obspy so that when new changes are proposed we can at least keep performance in the peripheral. Maybe someone has already done this?

**Comment by calum-chamberlain** [comment]
> Can you also try to pickle and un-pickle the same event on the same machine? I am curious how the un-pickling performance compares.

Not sure I know what you mean? That was un-pickling on the same machine for both instances, and I haven't changed how the pickler works, so haven't tested speeds there. Or do you just want to know how it compares for a single event rather than 1,000 events?

**Comment by d-chambers** [comment]
> Not sure I know what you mean?

Maybe I am the one who is confused. Let me back up: you mention two cases that you profiled. The first is current master the second is on the new branch. In the first method the catalog is stored on disk as quakeml, then parsed directory into a `Catalog` object. The second also stores the catalog on disk as quakeml but parses the quakeml to a dictionary, then the dictionary is converted to a `Catalog` object correct? If this is right, I am curious how reading the quakeml compares to reading a binary pickle dump from disk. It might help us roughly understand the performance cost of using xml as I imagine most of the object creation stuff would be about the same. 

So if you just did something like this in ipython:
```python
import pickle

import obspy

def unpickle_cat(path):
    with open(path, 'rb') as fi:
        return pickle.load(fi)

def pickle_cat(path):
    with open(path, 'wb') as fi:
         pickle.dump(cat, fi)

cat = obspy.read_events("the_big_catalog.xml")
pickle_cat('test.pkl')

%timeit unpickle_cat('test.pkl')
```
I am curious because awhile back I did some basic catalog IO profiling [here](https://gist.github.com/d-chambers/03088018405039a4acd81a1f995a256b) for the changes made in #2091 and I found that un-pickling the `pickle.dump` output for a smaller catalog was about 8X faster than reading it from QML, but I don't know if that will hold true for larger catalogs. Also, I found that using the `Catalog.copy` method was over 2X slower than using something like `pickle.loads(pickle.dumps(catalog))`, but I haven't had time to dig more into why this is.

**Comment by d-chambers** [comment]
Also, having a canonical dict representation of the `Catalog` object, and public functions for converting between `dict` and `Catalog`, would be extremely useful for at least two reasons:

1) It would be nice to have a format that works well with bigger catalogs. Eg, the 'Catalog' object gets a bit hard to work with somewhere around ~5,000 events. I imagine representing the same data with a nested dicts of basic data types would be much more scalable without the custom python object baggage (maybe ~500,000), although admittedly less convenient to work with.

2) It would make it much easier to use JSON as an input/output format, which can be a better choice than XML for some applications. I know @markcwill has already done some work on this in the `obspy.io.json` module so maybe he has some thoughts.  I also wonder if @krischer hasn't done something similar in Jane for storing events in a document format.

Anyway, I guess my point is that the intermediate dict representation you are using may be something useful in itself. Coming up with a single "right" way to represent a `Catalog` as a dictionary is probably a worthwhile discussion if it hasn't already been had.

**Comment by krischer** [comment]
I'm very much in favor of having a more efficient internal representation of `Catalog` objects. Instead of conversion functions we could also try to implement the `Catalog` object as a front-end to whatever the internal representation is. I feel like this might be easier to maintain in the long run and would naturally also speed up the `Catalog` object. I have not fully thought this through and it will probably be pretty tricky with all the various child objects but maybe it can be done.

We should also aim at having this usable for catalog sizes of ~1E6 events. Template matching approaches seem to yield these catalog sizes nowadays and being able to handle them with ObsPy would be great. Dictionaries might not really be up to the task but some in-memory relational SQLite tables or some kind of "relational" pandas tables could do it.

@calum-chamberlain 54 vs 241 seconds is already a great improvement! Do you know what the current bottleneck is? If you upload the code somewhere we could try to benchmark it. Splitting it up at the Event level and using the `multiprocessing` module should be very easy and it should give us an almost linear scaling with the number of cores (assuming pickling is fast or we have a more efficient internal representation).

**Comment by krischer** [comment]
In a similar vain I played around a bit with `etree.iterparse()` to extract coordinates from StationXML files in a much quicker way: https://github.com/SeismicData/pyasdf/blob/master/pyasdf/inventory_utils.py#L162

I don't see how to easily adapt this to our problem as we always want access to all children tags but maybe somebody has an idea.

**Comment by markcwill** [comment]
@d-chambers our JSON interface was specifically to create a GeoJSON schema'd dict from python objects with lat/lon coords (that could then be easily `json.dump`ed), but the function in `obspy.io.json` should work for encoding any `event` object and you could write its "decode" partner. I prefer `json.dumps(my_catalog, default=obspy.io.json.Default())` to something like `my_catalog.write(format='JSON')` but that's kind of personal preference I guess.

@calum-chamberlain & @krischer we have a serializer to QuakeML using the awesome `xmltodict` module, it uses the expat parser and is 1 file, so we vendor it into our qmlutil package. I have played around with deserializing, but haven't gotten a chance to really make it work yet. Rather than JSON, our goal was to de/serialize to python dict, with the hope that it would be pretty easy to construct Obspy classes from there. There are a couple gotchas but maybe there can be workarounds. The `xmltodict` module has pre- and post-processing hooks similar to json encode/decode default functions, so the "QuakeML" part could go in there... that's the thought we had. You can see our current serializer [here](https://github.com/NVSeismoLab/qmlutil/blob/master/qmlutil/xml.py), and a simple stub deserializer that just does basic types, no QuakeML-specific stuff yet.

The [xmltodict](https://github.com/martinblech/xmltodict) project uses expat, which supports streaming, but I'm not sure how fast it is compared to lxml, for example. If you have a file to use for benchmarking, point me to it? Or just for deserilizing to dict try:
`d = xmltodict.parse(open('query.quakeml.xml').read())`
...and see how it compares?

**Comment by calum-chamberlain** [comment]
So I ran some profiling, comparing my hack to obspy 1.1.0, and to raw pickling. The notebook is [here](https://gist.github.com/calum-chamberlain/5687bfab1890c7385dab4bb8e4fd10a6). I initially worked on a larger catalog, but ended up not using it all for the tests because it got slow waiting around, but the trends continue.

It looks like my custom xml to dictionary function is faster than xmltodict (by virtue of it not doing a lot of things, it doesn't interpret any of the values for example), but probably less safe.

Pickling is much faster than either implementation.

**Comment by d-chambers** [comment]
@jkmacc-LANL, good point. As long as we can mangle `obspy.core.Catalog` and friends into using an object that implements the `MutableMapping` interface as the "back-end" a lot of the scale issues could be solved. Doing that however, especially while maintaining the documented behaviour of `ResourceIdentifier`, could be very difficult.

**Comment by calum-chamberlain** [comment]
This has definitely gone over my head - @d-chambers and @jkmacc-LANL, can you point me in the direction of something to read that might help me understand those ideas, or should I just look at the examples @jkmacc-LANL posted above?

As an aside, I made some changes to my implementation to use multiprocessing, and it seems stable (hundreds of reads of 15,000 events, but the same file, so not really a good test), but not necessarily faster. I can do some more profiling, but it would be good to decide where we want to go with this before carrying on too much further I think. Happy to carry on working on it though as long as I'm not getting in the way too much! Also happy for someone else to play as well.

**Comment by dsentinel** [comment]
I don't use event catalogs much, so I may be missing some nuances, but this thread caught my eye as I enjoy optimization problems, and this OS project.

#### Reading XML
As @calum-chamberlain concluded the big time sink is `xpath()`, which fortunately isn't needed. Currently `xpath()` is being called, looped over, and then called on children looking for all needed tags. `xpath` is a SEARCH on an expression! And so it's searching the whole doc for all events, then searching for all the origins, then searching all mags , etc. Hitting many of the nodes in the DOM many times.
We want to create and object for each tag, and add it to the object tree. Sounds like iteration, and it is: Iterating over the DOM and creating, and adding the suitable object (or ignoring it, I assume there are some to be ignored as searching is being done with `xpath()`). The current implementation is iterating over a search, searching for 'nearly' everything. 
Iterating over each element is what `xmltodict` does, hence the speedup. Converting to a dict and then iterating over and creating the object tree will not be any faster than iterating over the xml, and creating a object tree directly.

Instead of:
```python
def _deserialize()
    for event_el in xpath():
        event = create_event(event_el)
        for origin_el in xpath():
            event.origins.append(create_origin(origin_el))
        for mag_el in xpath():
            event.magnitudes.append(create_magnitude(mag_el))
```
Something like:
```python
def _deserialize()
    for event_el in etree.iter('event'):
        event = create_event(event_el)
        for child in event_el.iterchildren():
            if child.tag == 'origin':
                event.origins.append(create_origin(child))
            elif child.tag == 'magnitude':
                event.magnitudes.append(create_mag(child))
```

#### Internal representation and processing format
This is a different discussion, perhaps for a different thread. XML is 'good' as an exchange format, but is not for a processing of larger data sets, and somewhere in between as a persistent store. 
What is the issue here, or is this just spun out of ideas solving the xml parsing slowness?
##### Internal rep
If the data is too big to fit in memory, or processing, in particular lookup, too slow? another back-end is in order.
##### persistent store
It sounds like pickle and zmap are currently working fast. Is performance here an issue?

**Comment by d-chambers** [comment]
Hey @calum-chamberlain,

Sorry for the late response.

> can you point me in the direction of something to read that might help me understand those ideas

The `MutableMapping` interface describes something that behaves like a dictionary.  It could be an actual dictionary, an external key-value store, an connection to a document database, etc. There are two important packages from the standard library in dealing with interfaces: The [`collections.abc` module ](https://docs.python.org/3/library/collections.abc.html) which describes a bunch of interfaces for basic data structures, and the [`abc`](https://docs.python.org/3/library/abc.html) package which provides the machinery to create your own interfaces/abstract base classes (but this should be done sparingly). If you have access to Fluent Python by Luciano Ramalho chapter 11 is an excellent primer on interfaces, duck/goose typing, and abstract base classes.

I started talking about the "back-end" of the catalog in order to address your second question: "Is there a faster IO alternative already implemented?" If generic `dict_to_catalog` and `catalog_to_dict` function existed this would open a lot of doors for easily implementing different methods for writing/reading catalog objects, many of which could be many times faster than xml.  

I am frantically working on a paper at the moment but I am very interested in helping out with this when I can, hopefully in a few weeks. 

@dsentinel, good point on the xpath! That does look promising and straight-forward to implement.

**Comment by jkmacc-LANL** [comment]
@calum-chamberlain Mad props for writing the first-pass dict implementation.  For expediency’s sake, have you compared it to the changes proposed by @dsentinel ?  If not, I might have a little time to try it...

**Comment by calum-chamberlain** [comment]
@jkmacc-LANL No I haven't had time to yet, if you have time it would be great to compare!

**Comment by jkmacc-LANL** [comment]
OK, I hope I did this right...  

Here are the baseline timings (from `master`), using the same `Catalog` of 1000+ events that @calum-chamberlain downloaded:
```python
In [1]: from obspy.io.quakeml.core import Unpickler

In [2]: for n_events in [1, 20, 100, 200]:
   ...:     print("Testing for {0} events".format(n_events))
   ...:     file_out = "test_catalog_{0}.xml".format(n_events)
   ...:     %timeit Unpickler().load(file_out)
   ...:
Testing for 1 events
213 ms ± 7.18 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Testing for 20 events
2.75 s ± 207 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Testing for 100 events
23.9 s ± 1.1 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
Testing for 200 events
38.6 s ± 2.96 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Here are the timings from [these changes](https://github.com/obspy/obspy/compare/QuakeML-iterchildren-reading), following @dsentinel 's suggestion:
```python
Testing for 1 events
209 ms ± 2.49 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Testing for 20 events
3.18 s ± 301 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Testing for 100 events
22 s ± 509 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Testing for 200 events
43.8 s ± 9.22 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

So, no real change.  I'm no XML guru, so someone who knows what they're doing should look this over.  `obspy-runtests io.quakeml` passes.

**Comment by calum-chamberlain** [comment]
It looks like you are still using `xpath` (via `self._xpath2obj` called by `self._pick`, `self._amplitude`, `self._magnitude`, `self._station_magnitude`) for a lot of the internal elements. I wonder if that is still the culprit?
You might be able to iterate through the children of the event tag, like @dsentinel suggested - checking that tag name with if statements if probably faster than using `xpath` lots. It might end up being quite a bit of work, maybe try changing how `self._picks` works for starters, which will cover quite a lot of elements in that catalog (ASFAIK) and see if that makes any difference, if it does, then hit the others?
The way I did it with the dictionary was essentially finding all of the calls to `self._xpath2obj` and getting rid of those, it's a big job, but those sections that do lots of looping (for instance, lots of calls usually come from picks) would be a useful place to test.

**Comment by jkmacc-LANL** [comment]
Geez, those things are everywhere.  Honestly, if it means doing as much work as you’ve already done, I’d kinda rather invest it in the dict version.  When I started, I didn’t realize @dsentinel ‘s post was pseudocode:-)

**Comment by calum-chamberlain** [comment]
@jkmacc-LANL they are indeed everywhere - I think if we did want to go down the `iterchildren` route it could be faster than my dict implementation (the one I did uses `getchildren` in a nested loop that is blind to the structure, and shamelessly). I don't having a go again (with `iterchildren` without the dictionary step), but not for a couple of weeks.

I realised I don't think I credited [this site](https://ericscrivner.me/2015/07/python-tip-convert-xml-tree-to-a-dictionary/), which is what the [`make_dict_from_tree`](https://github.com/obspy/obspy/blob/Quakeml-dictionary-reading/obspy/io/quakeml/core.py#L85) function is based on, with a couple of tweaks for this use.

I think the main thing is, we could probably go faster without the dictionary half-way house, but if we want to have Catalog objects based on a dictionary then maybe it makes sense to read to dict, then have some global dictionary-to-Catalog thing (which I think is what @d-chambers and @jkmacc-LANL were getting at? Thanks @d-chambers by the way, I will get up to speed on my abc's soon, ordered that book, and looking forward to becoming more fluent!)?

**Comment by dsentinel** [comment]
What are the use cases for processing catalogs? Could someone drop or point to some processing (psudeo)code?
If dicts are the way to go, what are the keys searched on? Are the searched, or iterated over? Filtered? on which keys?

**Comment by calum-chamberlain** [comment]
I imagine there are lots of use cases for using catalogs. Personally a common use-case I have is downloading waveforms for events and cutting them around picks (for templates in matched-filters), e.g. (psuedocode):
```python
streams = []
for event in catalog:
    bulk = []
    for pick in event.picks:
        # Items in tuple likely to be in the wrong order
        bulk.append((pick.waveform_id.station_code,
                               pick.waveform_id.network_code,
                               pick.waveform_id.location_code,
                               pick.waveform_id.channel_code,
                               pick.time - 10, pick.time + 90))
    st = client.get_waveforms_bulk(bulk)
    for tr in st:
        pick = lookuppick # not a thing at the moment
        tr.trim(pick.time - 0.5, pick.time + 5.5)
    streams.append(st)
return streams
```

I also often use the amplitudes to recalculate magnitudes based on different local magnitude-scales, and put my own amplitudes in there. I would have thought most attributes get hit by someone at some time. There are many possible use-cases. Not sure if giving any one or a few would be that useful? 

I also don't just iterate through things, e.g. the non-existent `lookuppick` in the pseudocode above would be to get a pick that matches the trace (based on `tr.id` and `pick.waveform_id`).

I also [filter catalogs based on location](https://github.com/eqcorrscan/EQcorrscan/blob/master/eqcorrscan/utils/catalog_utils.py#L166) and [picks based on frequency](https://github.com/eqcorrscan/EQcorrscan/blob/master/eqcorrscan/utils/catalog_utils.py#L28) and other things (which station or channel). Fairly sure that filtering could be done better.

**Comment by dsentinel** [comment]
Thanks @calum-chamberlain! I'll have a look.
Are there performance issues with catalogs scaling for such (once loaded)?
I've tested big inventories ~150MB, loading is painful, several minutes, but iterating/operating is reasonable.
It think before we start optimizing, we need to define what we're optimizing for.

**Comment by calum-chamberlain** [comment]
Agreed - I'm certainly pretty happy with working on Catalogs once loaded, I find it quite straightforward to get the parts I care about, and haven't found any noticeable speed or memory limitations once loaded. I'm sure they could be more memory efficient, but in most of the work that I do the Catalogs are smaller than other things (waveforms mostly).

The main issues are the speed of loading, and the memory consumption of writing large Catalogs.

I remain out of my depth with how a dictionary or abc backend to Catalog would work, so maybe someone else has something more useful to add.

**Comment by dsentinel** [comment]
> The main issues are the speed of loading, and the memory consumption of writing large Catalogs.

To clarify: When writing large catalogs to quakeml, memory usage is excessively high?

**Comment by calum-chamberlain** [comment]
@dsentinel, yes - the whole object is serialized in memory before dumping, which is probably the fastest way, but does use a lot of memory to keep both the object and a serialized object in memory. Happy to work around that on the user side though (e.g. don't try and write massive catalogs).

**Comment by d-chambers** [comment]
I agree with @calum-chamberlain about the IO speed being an issue, but I have yet to run into memory problems, although my workstation does have > 100 GB of ram. I don't have any issues with performance once the catalog is loaded into memory.

For example, I recently deployed 12 nodes over a mine for about a month. After running a STA/LTA and coincidence filter I detected ~ 12,000 events. I then stored the picks and nothing else in each `Event` object and saved the catalog to disk. It takes 2 minutes and 44 seconds to load the file with `obspy.read_events` from master, which is a significant bottleneck to the rest of my workflow.  

Having an intermediate dict format would be useful for a few reasons:

1) It would allow other serialization formats to simply convert the data to a dict, then there would be a well-tested `dict_to_catalog` method could be used rather than re-implementing all the object creation logic. This would be particularly useful for JSON or data stored in a json-like format in a database (eg MongoDB). 

2) It could enable swapping out the catalog backend, or making a lazy catalog, etc. This is something that would be far down the road and doesn't merit a lot of conversation in this thread. 

If it is a significant slowdown to use and intermediate dict, however, maybe it is not worth it.

**Comment by calum-chamberlain** [comment]
I do think the memory issue is likely to be *mostly* a non-issue for most people - its only been an issue for me for large catalogs (of a few thousand events) on my laptop with 16GB RAM, while I'm using memory for other things as well, e.g. when I'm being silly.

:+1: for having a useful `dict_to_catalog` and using JSON-like things.

**Comment by filefolder** [comment]
Sorry to revive this but the problem has gotten a lot worse in the last four years as ML pickers have made catalogs enormous. 

Would HDF5 be viable solution? I think there are ways to generate the proper hierarchy via a schema file but I have only started looking into it.  Seems like this idea is fairly popular in the adaptive meshing community. 

I note also that Seiscomp appears to be able to read large XML much much faster via C; to the extent that what I have to do now is import a catalog into seiscomp and then read it back in obspy via localhost FDSN.

**Comment by calum-chamberlain** [comment]
Just for reference, I think the seiscomp code is [here](https://github.com/SeisComP/common/tree/master/libs/seiscomp/io/xml). I don't know if/how HDF5 would solve the speed issue. From memory (and my memory is not great), I think the main issues that we had were in deserialising the xml doc. The actual reading of the xml was fairly fast. I don't know if deserialising from an HDF5 file would be faster without other changes.

**Comment by d-chambers** [comment]
In ObsPlus we defined [pydantic](https://pydantic-docs.helpmanual.io/) models which reflect ObsPy's catalog hierarchy. This enables fast translation to/from json and with future improvements to pydantic (the rust rewrite) it will only get faster, so you might give json a try. See the [docs](https://niosh-mining.github.io/obsplus/versions/latest/notebooks/utils/json.html)  for more details. 

My first impression is that HDF5 wont perform well if you mirror the hierarchical structure of quakeml exactly. If, however, you broke the structure down into a collection of tables it might work really well, but then why not just use [CSS](https://l2a.ucsd.edu/local/Manuals/CSS3.0_Format_Manual.pdf) and a proper database to begin with? [Pisces](https://github.com/LANL-Seismoacoustics/pisces) might help with this if you want to go the ORM route with this approach.

**Comment by filefolder** [comment]
Thanks both for the replies and links! I had not heard of ObsPlus but this looks like the easiest way forward. 

Just doing a quick comparison, loading 2621 events (picks, amplitudes, magnitudes, etc) took 2350 seconds via obspy and only **19 seconds** via ObsPlus / Json (having written to binary). Not sure how it scales at 10x or 100x but I would imagine as good or better. 

e.g.
```
	f = open('cat.bjson','rb')
	cat_json = f.readlines()
	cat_json = cat_json[0].decode() #back to a string
	f.close()
	cat = obsplus.json_to_cat(cat_json)
```

It seems like ObsPy integration of something similar to this was eventually abandoned (https://github.com/obspy/obspy/pull/2210), I can understand why, but it may be worth a rethink at some point.

**Comment by d-chambers** [comment]
@filefolder I am glad it works well for you.

>> It seems like ObsPy integration of something similar to this was eventually abandoned (https://github.com/obspy/obspy/pull/2210), I can understand why, but it may be worth a rethink at some point.

I agree with you. There are just two obstacles: 

1. Time: There is never enough of it
2. Dependencies: There are quite a few people using ObsPy who would rather keep a very short list of dependencies. This way it can run in more environments. To me, ObsPlus is a place for trying out "wouldn't it be cool if ObsPy ..." sorts of ideas that would require one or more significant dependencies (like pydantic, pandas, pytables, etc.).

**Comment by megies** [comment]
> what I have to do now is import a catalog into seiscomp and then read it back in obspy via localhost FDSN.

I'm confused. I must be missing something.. how would this be faster? If you request event data via FDSN thats just an additional step compared to just reading the data from a local file?

**Comment by filefolder** [comment]
> So, do those 19 secs reading time also include creating all the obspy objects in Python?

yes both times are from scratch to full access to the catalog. although I suppose saving the json version as a binary beforehand is cheating a little. I can be more rigorous with my testing next week.

> > what I have to do now is import a catalog into seiscomp and then read it back in obspy via localhost FDSN.
> 
> I'm confused. I must be missing something.. how would this be faster? If you request event data via FDSN thats just an additional step compared to just reading the data from a local file?

it does seem crazy but it only takes maybe 10 minutes to import a 10k+ catalog (scdb) and then just connect and download it back via FDSN to use in obspy. in contrast reading directly via obspy is usually a “let it run overnight” operation.

**Comment by megies** [comment]
> it does seem crazy but it only takes maybe 10 minutes to import a 10k+ catalog (scdb) and then just connect and download it back via FDSN to use in obspy. in contrast reading directly via obspy is usually a “let it run overnight” operation.

I still dont get it. Makes no sense to me.

**Comment by d-chambers** [comment]
I think @dsentinel said the xml's parser's poor performance and scaling is [due to xpath](https://github.com/obspy/obspy/issues/1910#issuecomment-411495314) searching the whole xml document for certain tags for each event.

**Comment by megies** [comment]
> I think @dsentinel said the xml's parser's poor performance and scaling is [due to xpath](https://github.com/obspy/obspy/issues/1910#issuecomment-411495314) searching the whole xml document for certain tags for each event.

that could explain how obsplus is faster on a plain read, although its hardly surprising if you read from some other format (some binary or json or something?) where you expect everything to be in the exact place, which will just fail if anything is out of place, as opposed to xml where everything might be all over the place and the reading is robust.

If obsplus is still parsing xml and free form, and not in some fully compiled code, let us know.

This is just about use cases, to be honest. Most users read some small quakeml and it should be robust. And then some power users want to read 5k events with tons of picks each and ideally within seconds. It's just two worlds, really.

But still doest'n explain how an obspy FDSNWS client request should be faster than a simple local obspy `read_inventory`, still wanna know what thats about

**Comment by d-chambers** [comment]
The ObsPlus parser uses [pydantic](https://pydantic-docs.helpmanual.io/) to read json, not xml. All the compiled magic is happening in pydantic. 

I didn't realize the xml elements could be *anywhere* in the document, I guess that's why we use xpath even though it scales poorly. I wonder if we could instead parse all the elements into a hash table then just do fast look ups on ids for reconstructing the hierarchy?

**Comment by megies** [comment]
> I wonder if we could instead parse all the elements into a hash table then just do fast look ups on ids for reconstructing the hierarchy?

sounds like it could help, but sounds like quite some work too.
